hathor/merged_mining/coordinator.py

```diff
@@ -530,21 +534,37 @@ def handle_configure(self, params: List, msgid: Optional[str]) -> None:
 
         self.send_result(res, msgid)
 
-    def handle_subscribe(self, params: Any, msgid: Optional[str]) -> None:
+    def handle_subscribe(self, params: List[str], msgid: Optional[str]) -> None:
         """ Handles subscribe request by answering it and triggering a job request.
 
         :param msgid: JSON-RPC 2.0 message id
         :type msgid: Optional[str]
         """
+        from math import log2
         assert self.miner_id is not None
         self._subscribed = True
         self.subscribed_at = time.time()
         self.log.info('Miner subscribed', address=self.miner_address)
         # session = str(self.miner_id)
         session = [['mining.set_difficulty', '1'], ['mining.notify', str(self.miner_id)]]
         self.send_result([session, self.xnonce1.hex(), self.xnonce2_size], msgid)
+        self.user_agent = params[0]
+        if 'nicehash' in self.user_agent.lower():
+            self.log.info('special case mindiff for NiceHash')
+            self.initial_difficulty = PDiff(NICEHASH_MIN_DIFF)
+            self.min_difficulty = NICEHASH_MIN_DIFF
+        if '/' in self.user_agent:
+            try:
+                # example: bmminer/2.0.0/Antminer S9j/14500
+                # the last part will often contain an indication of the hashrate (in GH/s)
+                hashrate_ghs = int(self.user_agent.split('/')[-1])
+                self.log.debug('detected hashrate', hashrate_ghs=hashrate_ghs)
+            except ValueError:
+                pass
+            else:
+                self.initial_difficulty = Weight(log2(hashrate_ghs * self.TARGET_JOB_TIME) + log2(1e9)).to_pdiff()
```

Comment:
Do we need the `else`, can't this just run inside the `try` block?

Reply:
I'd rather not accidentally catch any exceptions that may come from this, because any exceptions here would actually be an error.

Reply:
Got it.